File: bfs/src/bfs.rs
use crate::{add_to_open::add_to_open, print_map::MainState, Pos, COLS, ROWS};
use ggez::Context;
use std::{collections::LinkedList, thread::sleep, time::Duration};

pub fn bfs(main_state: &mut MainState, ctx: &mut Context) {
    let mut open: LinkedList<Pos> = LinkedList::new();
    let mut closed: LinkedList<Pos> = LinkedList::new();

    open.push_back(main_state.start.clone());

    while let Some(current) = open.pop_front() {
        println!("X: {}, Y: {}", current.x, current.y);

        main_state.map[current.y as usize][current.x as usize] = "+";

        sleep(Duration::from_millis(10));

        main_state.print_map(ctx).expect("Erro ao desenhar o mapa.");

        if current.x == main_state.goal.x && current.y == main_state.goal.y {
            println!("Destino encontrado!");

            let mut cursor = Some(current);

            while let Some(pos) = cursor {
                if pos == main_state.start || pos == main_state.goal {
                    break;
                }

                main_state.map[pos.y as usize][pos.x as usize] = "*";

                cursor = pos.last.as_deref().cloned();

                main_state.print_map(ctx).expect("Erro ao desenhar o mapa.");
            }

            for row in main_state.map.iter_mut() {
                for cell in row.iter_mut() {
                    if *cell == "+" {
                        *cell = " ";
                    }
                }
            }

            main_state.print_map(ctx).expect("Erro no desenho do mapa.");
            return;
        }

        closed.push_back(current.clone());

        if current.y > 0 && main_state.map[current.y as usize][current.x as usize] == " " {
            add_to_open(
                main_state.map,
                &mut open,
                &mut closed,
                &current,
                current.x,
                current.y - 1,
            );
        }

        if current.y < (ROWS - 1) as u32 {
            add_to_open(
                main_state.map,
                &mut open,
                &mut closed,
                &current,
                current.x,
                current.y + 1,
            );
        }

        if current.x > 0 {
            add_to_open(
                main_state.map,
                &mut open,
                &mut closed,
                &current,
                current.x - 1,
                current.y,
            );
        }

        if current.x < (COLS - 1) as u32 {
            add_to_open(
                main_state.map,
                &mut open,
                &mut closed,
                &current,
                current.x + 1,
                current.y,
            );
        }
    }

    println!("Destino não encontrado!");
}

File: bfs/java_bfs/src/BFS.java
import java.util.LinkedList;

public class BFS {

    private static final int ROWS = Main.ROWS;
    private static final int COLS = Main.COLS;


    public static void bfs(char[][] map, Pos start, Pos goal) {

        LinkedList<Pos> open = new LinkedList<Pos>();

        open.add(start);

        while (!open.isEmpty()) {
            Pos current = open.pop();

            Main.clearScreen();
            Main.printMap(map);

            try {
                Thread.sleep(10);
            } catch (InterruptedException e) {
                System.out.println(e.getMessage());
            }

            if (current.x == goal.x && current.y == goal.y) {
                System.out.println("Destino encontrado!");
                Main.clear_map(map);
                Main.backtrack(map, current);
                return;
            }


            if (current.y > 0 && (map[current.y - 1][current.x] == ' ' || map[current.y - 1][current.x] == 'G')) {
                Pos dummy = new Pos(current.y - 1, current.x, current);
                check_lists(map, open, dummy);
            }

            if (current.y < ROWS - 1 && (map[current.y + 1][current.x] == ' ' || map[current.y + 1][current.x] == 'G')) {
                Pos dummy = new Pos(current.y + 1, current.x, current);
                check_lists(map, open, dummy);
            }

            if (current.x > 0 && (map[current.y][current.x - 1] == ' ' || map[current.y][current.x - 1] == 'G')) {
                Pos dummy = new Pos(current.y, current.x - 1, current);
                check_lists(map, open, dummy);
            }

            if (current.x < COLS - 1 && (map[current.y][current.x + 1] == ' ' || map[current.y][current.x + 1] == 'G')) {
                Pos dummy = new Pos(current.y, current.x + 1, current);
                check_lists(map, open, dummy);
            }
        }
    }

    public static void check_lists(char[][] map, LinkedList<Pos> open, Pos current) {
        map[current.y][current.x] = '+';
        open.add(current);
    }


}
